strawberryfields/apps/graph/dense.py

```diff
@@ -0,0 +1,179 @@
+
+def find_dense(
+    graph: graph_type, nodes: int, iterations: int = 1, options: Optional[dict] = None
+) -> Tuple[float, list]:
+    """Find a dense subgraph of a given size.
+
+    This function returns the densest `node-induced subgraph
+    <http://mathworld.wolfram.com/Vertex-InducedSubgraph.html>`__ of size ``nodes`` after
+    multiple repetitions. It uses heuristic optimization that combines search space exploration
+    with local searching. The heuristic method can be set with the ``options`` argument. Methods
+    can contain stochastic elements, where randomness can come from distributions including GBS.
+
+    All elements of the heuristic can be controlled with the ``options`` argument, which should be a
+    dict-of-dicts where the first level specifies the option type as a string-based key,
+    with corresponding value being a dictionary of options for that type. The option types are:
+
+    - ``"heuristic"``: specifying options used by optimization heuristic; corresponding
+      dictionary of options explained further :ref:`below <heuristic>`
+    - ``"backend"``: specifying options used by backend quantum samplers; corresponding
+      dictionary of options explained further in :mod:`~strawberryfields.future.apps.sample`
+    - ``"resize"``: specifying options used by resizing method; corresponding dictionary of
+      options explained further in :mod:`~strawberryfields.future.apps.graph.resize`
+    - ``"sample"``: specifying options used in sampling; corresponding dictionary of options
+      explained further in :mod:`~strawberryfields.future.apps.graph.sample`
+
+    If unspecified, a default set of options is adopted for a given option type.
+
+    .. _heuristic:
+
+    The options dictionary corresponding to ``"heuristic"`` can contain any of the following:
+
+    .. glossary::
+
+        key: ``"method"``, value: *str* or *callable*
+            Value can be either a string selecting from a range of available methods or a
+            customized callable function. Options include:
+
+            - ``"random-search"``: a simple random search algorithm where many subgraphs are
+              selected and the densest one is chosen (default)
+            - *callable*: accepting ``(graph: nx.Graph, nodes: int, iterations: int, options:
+              dict)`` as arguments and returning ``Tuple[float, list]`` corresponding to the
+              density and list of nodes of the densest subgraph found, see :func:`random_search`
+              for an example
+
+    Args:
+        graph (graph_type): the input graph
+        nodes (int): the size of desired dense subgraph
+        iterations (int): number of iterations to use in algorithm
+        options (dict[str, dict[str, Any]]): dict-of-dicts specifying options in different parts
+            of heuristic search algorithm; defaults to :const:`OPTIONS_DEFAULTS`
+
+    Returns:
+        Tuple[float, list]: the density and list of nodes corresponding to the densest subgraph
```

Comment:
should tuple be capitalized here? When used directly from the `Typing` module yes (to avoid confusion with the inbuilt `tuple()`), but as a Sphinx return-type, my understanding was it should be `tuple`
```suggestion
        tuple[float, list]: the density and list of nodes corresponding to the densest subgraph
```

Reply:
I think Sphinx shouldn't generate a link here either way, so it's just for the benefit of the human readers. Technically we're already using the `typing` generic syntax when we write e.g. `tuple[float]`. OTOH, so far we have not used the `Tuple[Union[x, y]]` syntax suggested by `typing`, instead we tend to write `tuple[x, y]` because it's easier and maybe a bit more readable...
Adding to the confusion, I myself have been occasionally using `dict[x->y]`, which in `typing` notation would be `Dict[x, y]`.

I think this really only matters if we started using Python 3 type annotations instead of specifying the argument types in the docstring.

Reply:
Thanks @smite!